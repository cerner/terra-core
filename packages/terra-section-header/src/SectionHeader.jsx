import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import KeyCode from 'keycode-js';
import Arrange from 'terra-arrange';
import styles from './SectionHeader.module.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * Text to be displayed on the SectionHeader.
   */
  title: PropTypes.string.isRequired,
  /**
   * Callback function triggered when the accordion icon is clicked.
   */
  onClick: PropTypes.func,
  /**
   * Whether the accordion icon should be displayed in its 'open' or 'closed' position.
   */
  isOpen: PropTypes.bool,
  /**
   * Optionally sets the heading level. One of `1`, `2`, `3`, `4`, `5`, `6`. Default `level=2`.
   */
  level: PropTypes.oneOf([1, 2, 3, 4, 5, 6]),
  /**
   * Sets the background of the section header to transparent.
   */
  isTransparent: PropTypes.bool,
};

const defaultProps = {
  onClick: undefined,
  isOpen: false,
  isTransparent: false,
  level: 2,
};

const isRecognizedKeyPress = event => ((event.nativeEvent.keyCode === KeyCode.KEY_RETURN) || (event.nativeEvent.keyCode === KeyCode.KEY_SPACE));

class SectionHeader extends React.Component {
  constructor(props) {
    super(props);

    this.state = { isActive: false };
    this.wrapOnKeyDown = this.wrapOnKeyDown.bind(this);
    this.wrapOnKeyUp = this.wrapOnKeyUp.bind(this);
  }

  wrapOnKeyDown(onKeyDown) {
    return ((event) => {
      if (isRecognizedKeyPress(event) && !this.state.isActive) {
        this.setState({ isActive: true });

        // Call the same function the user provides for a click event
        if (this.props.onClick) {
          this.props.onClick(event);
        }
      }

      if (onKeyDown) {
        onKeyDown(event);
      }
    });
  }

  wrapOnKeyUp(onKeyUp) {
    return ((event) => {
      if (isRecognizedKeyPress(event) && this.state.isActive) {
        this.setState({ isActive: false });
      }

      if (onKeyUp) {
        onKeyUp(event);
      }
    });
  }

  render() {
    const {
      title,
      onClick,
      isOpen,
      isTransparent,
      level,
      ...customProps
    } = this.props;

    if ((process.env.NODE_ENV !== 'production') && (!onClick && isOpen)) {
      // eslint-disable-next-line no-console
      console.warn('\'isOpen\' are intended to be used only when \'onClick\' is provided.');
    }

    const attributes = Object.assign({}, customProps);

    if (onClick) {
      attributes.tabIndex = '0';
      attributes.onKeyDown = this.wrapOnKeyDown(attributes.onKeyDown);
      attributes.onKeyUp = this.wrapOnKeyUp(attributes.onKeyUp);
      attributes.role = 'button';
    }

    const iconClassNames = cx([
      'accordion-icon',
      { 'is-open': onClick && isOpen },
    ]);

    const accordionIcon = (
      <div className={cx('accordion-icon-wrapper')}>
        <span className={iconClassNames} />
      </div>
    );

    const sectionHeaderClassNames = cx([
      'section-header',
      { 'is-interactable': onClick },
      { 'is-active': this.state.isActive },
      { 'is-transparent': isTransparent },
      customProps.className,
    ]);

    const Element = `h${level}`;

    // allows us to set an onClick on the div
    // We set key events and role conditionally set if onClick is set
    // eslint doesn't know about this and so it marks this as a lint error
    /* eslint-disable jsx-a11y/click-events-have-key-events */
    /* eslint-disable jsx-a11y/no-static-element-interactions */
    return (
      <div {...attributes} onClick={onClick} className={sectionHeaderClassNames}>
        <Arrange
          fitStart={onClick && accordionIcon}
          fill={<Element className={cx('title')}>{title}</Element>}
        />
      </div>
    );
    /* eslint-enable jsx-a11y/no-static-element-interactions */
  }
}

SectionHeader.propTypes = propTypes;
SectionHeader.defaultProps = defaultProps;

export default SectionHeader;
