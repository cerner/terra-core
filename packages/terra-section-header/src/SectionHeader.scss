:local {
  .section-header {
    /* TODO: Add styles */
  }
}
