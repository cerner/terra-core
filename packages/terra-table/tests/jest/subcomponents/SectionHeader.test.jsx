import React from 'react';
import SectionHeader from '../../../src/subcomponents/_SectionHeader';

// Snapshot Tests
it('should render default', () => {
  const shallowComponent = shallow(<SectionHeader title="test" numberOfColumns={1} />);
  expect(shallowComponent).toMatchSnapshot();
});

it('should render with isCollapsed', () => {
  const shallowComponent = shallow(<SectionHeader title="test" isCollapsed numberOfColumns={1} />);
  expect(shallowComponent).toMatchSnapshot();
});

it('should render with isCollapsible', () => {
  const shallowComponent = shallow(<SectionHeader title="test" isCollapsible numberOfColumns={1} />);
  expect(shallowComponent).toMatchSnapshot();
});

it('should render with callback functions', () => {
  const mockCallBack = jest.fn();

  const shallowComponent = shallow(
    <SectionHeader title="test" isCollapsible onSelect={mockCallBack} refCallback={jest.fn()} numberOfColumns={1} />,
  );
  expect(shallowComponent).toMatchSnapshot();
  shallowComponent.find('[role="columnheader"]').simulate('click');
  shallowComponent.find('[role="columnheader"]').simulate('keydown', { nativeEvent: { keyCode: 13 } });
  shallowComponent.find('[role="columnheader"]').simulate('keydown', { nativeEvent: { keyCode: 32 } });
  expect(mockCallBack.mock.calls.length).toEqual(3);
});
