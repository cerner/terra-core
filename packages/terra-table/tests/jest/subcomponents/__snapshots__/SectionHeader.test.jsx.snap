// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render default 1`] = `
<div
  className="section-header"
  role="row"
>
  <div
    className="section-content"
    colSpan={1}
    headers=""
    role="columnheader"
  >
    <span
      className="title"
    >
      test
    </span>
  </div>
</div>
`;

exports[`should render with callback functions 1`] = `
<div
  className="section-header is-collapsible"
  role="row"
>
  <div
    aria-expanded={true}
    className="section-content"
    colSpan={1}
    data-row-show-focus="true"
    headers=""
    onBlur={[Function]}
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    role="columnheader"
    tabIndex="0"
  >
    <div
      aria-hidden={true}
      className="start"
    >
      <span
        className="accordion-icon is-open"
      />
    </div>
    <div
      className="fill"
    >
      <span
        className="title"
      >
        test
      </span>
    </div>
  </div>
</div>
`;

exports[`should render with isCollapsed 1`] = `
<div
  className="section-header"
  role="row"
>
  <div
    className="section-content"
    colSpan={1}
    headers=""
    role="columnheader"
  >
    <span
      className="title"
    >
      test
    </span>
  </div>
</div>
`;

exports[`should render with isCollapsible 1`] = `
<div
  className="section-header is-collapsible"
  role="row"
>
  <div
    aria-expanded={true}
    className="section-content"
    colSpan={1}
    data-row-show-focus="true"
    headers=""
    onBlur={[Function]}
    onMouseDown={[Function]}
    role="columnheader"
    tabIndex="0"
  >
    <div
      aria-hidden={true}
      className="start"
    >
      <span
        className="accordion-icon is-open"
      />
    </div>
    <div
      className="fill"
    >
      <span
        className="title"
      >
        test
      </span>
    </div>
  </div>
</div>
`;
